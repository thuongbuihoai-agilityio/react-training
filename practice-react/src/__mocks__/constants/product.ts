import { Product } from "@common-types/product";
import iceCreamSundae from "@assets/images/ice-cream-sundae.png";

const PRODUCT_MOCKING: Product = {
  productId: "16466545464646547",
  categoryId: "1651454657522",
  productName: "Ice Cream Sundae",
  productDescription:
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
  productQuantity: 5,
  originalPrice: {
    value: 20,
    currency: "$",
  },
  discountPrice: {
    value: 20,
    currency: "$",
  },
  offerQuantity: "3kg",
  offer: false,
  bestSelling: true,
  popular: false,
  images: {
    src: iceCreamSundae,
    alt: "This is ice-cream-sundae images",
  },
};
const PRODUCT_MOCKING_LIST: Product[] = [
  {
    productId: "16466545464646547",
    categoryId: "1651454657522",
    productName: "Ice Cream Sundae",
    productDescription:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
    productQuantity: 5,
    originalPrice: {
      value: 20,
      currency: "$",
    },
    discountPrice: {
      value: 20,
      currency: "$",
    },
    offerQuantity: "3kg",
    offer: false,
    bestSelling: true,
    popular: false,
    images: {
      src: "/src/assets/images/ice-cream-sundae.png",
      alt: "This is ice-cream-sundae images",
    },
  },
  {
    productId: "1646654668992646547",
    categoryId: "1651924356334",
    productName: "Beef-Steak",
    productDescription:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam molestie ultrices donec pulvinar quis est bibendum suspendisse viverra.",
    productQuantity: 9,
    originalPrice: {
      value: 50,
      currency: "$",
    },
    discountPrice: {
      value: 50,
      currency: "$",
    },
    offerQuantity: "5kg",
    offer: false,
    bestSelling: false,
    popular: true,
    images: {
      src: "/src/assets/images/beef-steak.png",
      alt: "This is beef-steak images",
    },
  },
];

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
