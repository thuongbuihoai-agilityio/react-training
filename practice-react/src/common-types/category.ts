export interface Categories {
  id: string;
  name: string;
  images: {
    src: string;
    alt: string;
  };
}

export interface CategoryContext {
  searchValue: {} | undefined;
  setSearchValue: Function;
  categories: Categories[];
  setCategories: Function;
}