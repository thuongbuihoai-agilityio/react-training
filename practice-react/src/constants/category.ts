import { CategoryType } from "@common-types/categoryList";
import iconBeverages from "@assets/images/icons/beverages.png";
import iconDairy from "@assets/images/icons/dairy.png";
import iconMeat from "@assets/images/icons/meat.png";
import iconSpicesHerbs from "@assets/images/icons/spices-herbs.png";
import iconStationery from "@assets/images/icons/stationery.png";
import iconCostmetic from "@assets/images/icons/costmetic.png";

const CATEGORY: CategoryType = {
  key: "beverages",
  src: iconBeverages,
  text: "Beverages",
  alt: "This is icon beverages",
};

const CATEGORY_LIST: CategoryType[] = [
  {
    key: "beverages",
    src: iconBeverages,
    text: "Beverages",
    alt: "This is icon beverages",
  },
  {
    key: "dairy",
    src: iconDairy,
    text: "Dairy",
    alt: "This is icon dairy",
  },
  {
    key: "meat",
    src: iconMeat,
    text: "Meat",
    alt: "This is icon meat",
  },
  {
    key: "spices-herbs",
    src: iconSpicesHerbs,
    text: "Spices & Herbs",
    alt: "This is icon spices-herbs",
  },
  {
    key: "stationery",
    src: iconStationery,
    text: "Stationery",
    alt: "This is icon stationery",
  },
  {
    key: "costmetic",
    src: iconCostmetic,
    text: "Costmetic",
    alt: "This is icon costmetic",
  },
];

export { CATEGORY, CATEGORY_LIST };
