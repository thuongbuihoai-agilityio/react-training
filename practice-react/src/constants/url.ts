const BASE_URL = "http://localhost:1234";
const PRODUCTS_URL = `${BASE_URL}/products`;
const CATEGORIES_URL = `${BASE_URL}/categories`;

export { BASE_URL, PRODUCTS_URL, CATEGORIES_URL };