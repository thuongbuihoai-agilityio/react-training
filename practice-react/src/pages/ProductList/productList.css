.productList {
  display: flex;
}

.productList__page {
  margin: 100px 0 70px 443px;
  font-size: 24px;
  line-height: 36px;
  color: #979797;
}

.productList__select {
  background: #F4F1E9;
}

.productList__select--category {
  padding: 32px 198px 47px 98px;
  font-weight: 600;
  font-size: 27px;
  line-height: 40px;
  color: #2B463C;
}

.productList--popular {
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 0 99px;
}