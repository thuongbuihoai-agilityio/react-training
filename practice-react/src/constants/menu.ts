import { MenuType } from "../common-types/menu";

const MENU_LIST: MenuType[] = [
  {
    key: "orders",
    label: "Orders",
  },
  {
    key: "cart",
    label: "Cart",
  },
  {
    key: "contact",
    label: "Contact",
  },
];

const MENU_SERVICE: MenuType[] = [
  {
    key: "email-marketing",
    label: "Email Marketing",
  },
  {
    key: "campaigns",
    label: "Campaigns",
  },
  {
    key: "branding",
    label: "Branding",
  },
  {
    key: "offline",
    label: "Offline",
  },
];

const MENU_ABOUT: MenuType[] = [
  {
    key: "our-story",
    label: "Our Story",
  },
  {
    key: "benefits",
    label: "Benefits",
  },
  {
    key: "team",
    label: "Team",
  },
  {
    key: "careers",
    label: "Careers",
  },
];

const MENU_HELP: MenuType[] = [
  {
    key: "FAQs",
    label: "FAQs",
  },
  {
    key: "contact-us",
    label: "Contact Us",
  },
];

export { MENU_LIST, MENU_SERVICE, MENU_ABOUT, MENU_HELP };
