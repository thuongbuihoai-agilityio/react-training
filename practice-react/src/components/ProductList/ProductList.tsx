import React from "react";
import { Link } from "react-router-dom";
import { Product } from "@common-types/product";
import Text from "@components/common/Text/Text";
import Price from "@components/common/Price/Price";
import Counter from "@components/common/Counter/Counter";
import Icon from "@components/common/Icon/Icon";
import "./productList.css";

interface CartProductProps {
  type: string;
  content: string;
  visibleQuantity?: boolean;
  visibleDiscountPrice?: boolean;
  visibleCounter?: boolean;
  productList?: Product[];
}

const CardProductList: React.FC<CartProductProps> = ({
  type="normal",
  content="discount",
  visibleQuantity,
  visibleDiscountPrice,
  visibleCounter,
  productList,
}) => {
  let className = "product-list";
  switch (type) {
    case "normal":
      className = `${className}-${type}`;
      break;
    case "medium":
      className = `${className}-${type}`;
      break;
    case "medium-box":
      className = `${className}-${type}`;
      break;
    default:
      throw new Error("Invalid type");
  }

  let cartInfo = "card";
  switch (content) {
    case "discount":
      cartInfo = `${cartInfo}-${content}`;
      break;
    case "quantity":
      cartInfo = `${cartInfo}-${content}`;
      break;
    default:
      throw new Error("Invalid content");
  }

  return (
    <>
      {productList?.map((product: Product) => (
        <div
          data-testid="product-list"
          key={product.productId}
          className={className}
        >
          <img
            className="card-image"
            src={product.images.src}
            alt={product.images.alt}
          />
          <div className={cartInfo}>
            <Link className="cart-link" to={`/products/${product?.productId}`}>
              <Text size="normal" decoration="" text={product?.productName} />
            </Link>
            {visibleQuantity && (
              <p className="card-unit">{product?.offerQuantity}</p>
            )}
            <Price
              type="original"
              value={product?.originalPrice.value}
              currency={product?.originalPrice.currency}
            />
            {visibleDiscountPrice && (
              <Price
                type="discount"
                value={product?.discountPrice.value}
                currency={product?.discountPrice.currency}
              />
            )}
          </div>
          {visibleCounter && (
            <div className="card-counter">
              <Counter counter={product.productQuantity} />
              <div className="card-icon">
                <Icon iconName="cart" />
              </div>
            </div>
          )}
        </div>
      ))}
    </>
  );
};

export default CardProductList;
