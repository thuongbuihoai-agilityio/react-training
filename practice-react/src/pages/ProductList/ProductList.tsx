import React, { memo, useCallback, useContext, useMemo, useState } from "react";
import { DataContext } from "@context/DataContext";
import { useParams } from "react-router-dom";
import CategoryList from "@components/Category/CategoryList/CategoryList";
import NavigationBar from "@components/common/NavigationBar/NavigationBar";
import CardProductList from "@components/CardProduct/CardProductList";
import "./productList.css";

const ProductList: React.FC = memo(() => {
  const { products } = useContext(DataContext);
  const { id } = useParams();

  const defaultCategories = useMemo(() => {
    return id ? [id] : [];
  }, [id]);

  const [selectedCategories, setSelectedCategories] = useState<string[]>(defaultCategories);
  const productList = useMemo(() => {
    return selectedCategories.length
      ? products.filter((product) =>
          selectedCategories.includes(product.categoryId)
        )
      : products;
  }, [products, selectedCategories]);

  const handleToggleCategory = useCallback(
    (categoryId: string, value: string) => {
      if (categoryId) {
        if (value) {
          setSelectedCategories([...selectedCategories, categoryId]);
        } else {
          setSelectedCategories(
            selectedCategories.filter((item) => item !== categoryId)
          );
        }
      } else {
        if (value) {
          setSelectedCategories([]);
        }
      }
    },
    [selectedCategories]
  );

  return (
    <div data-testid="product-list">
      <NavigationBar isThemeDark={true} />
      <p className="productList__page">Home/Beverages</p>
      <div className="productList">
        <div className="productList__select">
          <p className="productList__select--category">Categories</p>
          <CategoryList
            id={id}
            type="checkbox"
            isCheckbox={true}
            selectedCategories={selectedCategories}
            onToggleCategory={handleToggleCategory}
          />
        </div>
        <div className="productList--popular">
          <CardProductList
            type="popular"
            content="popular"
            visibleCounter={true}
            productList={productList}
          />
        </div>
      </div>
    </div>
  );
});

export default ProductList;
