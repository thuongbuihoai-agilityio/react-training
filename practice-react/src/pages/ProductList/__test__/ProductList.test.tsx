import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { DataContextProps } from "@common-types/data";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/category";
import { PRODUCT_MOCKING_LIST } from "@__mocks__/constants/product";
import { DataContext } from "@context/DataContext";
import "@testing-library/jest-dom";
import ProductList from "../ProductList";

const contextProductMock: DataContextProps = {
  searchValue: {},
  setSearchValue: jest.fn(),
  categories: CATEGORY_MOCKING_LIST,
  setCategories: jest.fn(),
  products: PRODUCT_MOCKING_LIST,
  setProducts: jest.fn(),
};

describe("ProductList component", () => {
  test("should render ProductList component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductList />
        </Router>
      </DataContext.Provider>
    );
    expect(getByTestId("product-list")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductList />
        </Router>
      </DataContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
