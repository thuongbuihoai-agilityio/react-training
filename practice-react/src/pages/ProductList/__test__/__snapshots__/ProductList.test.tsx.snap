// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="product-list"
  >
    <nav
      class="nav"
      data-testid="navigation-bar"
    >
      <a
        href="/"
      />
      <a
        data-testid="logo"
        href="/"
      >
        <img
          alt="This is logo"
          class=""
        />
      </a>
      <div
        class="nav__about"
      >
        <div
          class="menu"
          data-testid="menu"
        >
          <ul
            class="menu__list"
          >
            <li
              class="menu__item"
              data-testid="menu-item"
            >
              <a
                href="/products"
              >
                Shop
              </a>
            </li>
            <li
              class="menu__item"
            >
              <a
                href=""
              >
                Orders
              </a>
            </li>
            <li
              class="menu__item"
            >
              <a
                href=""
              >
                Cart
              </a>
            </li>
            <li
              class="menu__item"
            >
              <a
                href=""
              >
                Contact
              </a>
            </li>
          </ul>
        </div>
        <button
          class="btn btn--dark"
        >
          Sign Up
        </button>
        <button
          class="btn btn__outline--dark"
        >
          Login
        </button>
      </div>
    </nav>
    <p
      class="productList__page"
    >
      Home/Beverages
    </p>
    <div
      class="productList"
    >
      <div
        class="productList__select"
      >
        <p
          class="productList__select--category"
        >
          Categories
        </p>
        <div
          class="categoryList--checkbox"
          data-testid="category-list"
        >
          <div
            class="categoryList__checkbox"
          >
            <label
              class="checkbox__value"
            >
              <input
                data-index=""
                multiple=""
                name="checkbox"
                type="checkbox"
              />
              All
            </label>
          </div>
          <div>
            <div
              class="categoryList__checkbox"
            >
              <label
                class="checkbox__value"
              >
                <input
                  data-index="1234"
                  multiple=""
                  name="checkbox"
                  type="checkbox"
                />
                Beverages
              </label>
            </div>
          </div>
          <div>
            <div
              class="categoryList__checkbox"
            >
              <label
                class="checkbox__value"
              >
                <input
                  data-index="12345"
                  multiple=""
                  name="checkbox"
                  type="checkbox"
                />
                Dairy
              </label>
            </div>
          </div>
          <div>
            <div
              class="categoryList__checkbox"
            >
              <label
                class="checkbox__value"
              >
                <input
                  data-index="12345"
                  multiple=""
                  name="checkbox"
                  type="checkbox"
                />
                Meat
              </label>
            </div>
          </div>
          <div>
            <div
              class="categoryList__checkbox"
            >
              <label
                class="checkbox__value"
              >
                <input
                  data-index="12345"
                  multiple=""
                  name="checkbox"
                  type="checkbox"
                />
                Spices & Herbs
              </label>
            </div>
          </div>
          <div>
            <div
              class="categoryList__checkbox"
            >
              <label
                class="checkbox__value"
              >
                <input
                  data-index="12345"
                  multiple=""
                  name="checkbox"
                  type="checkbox"
                />
                Stationery
              </label>
            </div>
          </div>
          <div>
            <div
              class="categoryList__checkbox"
            >
              <label
                class="checkbox__value"
              >
                <input
                  data-index="12345"
                  multiple=""
                  name="checkbox"
                  type="checkbox"
                />
                Costmetic
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="productList--popular"
      >
        <div
          class="card__product--popular"
          data-testid="card-product-list"
        >
          <img
            alt="This is ice-cream-sundae images"
            class="card__image"
            src="/src/assets/images/ice-cream-sundae.png"
          />
          <div
            class="card__popular"
          >
            <a
              class="cart__link"
              href="/products/16466545464646547"
            >
              <p
                class="text--normal"
                data-testid="text"
              >
                Ice Cream Sundae
              </p>
            </a>
            <p
              class="price__original"
              data-testid="price"
            >
              <span
                class="price--unit"
              >
                $
              </span>
              20
            </p>
          </div>
          <div
            class="card__counter"
          >
            <div
              class="counter"
              data-testid="counter"
            >
              <i
                class="fa fa-minus"
              />
              <input
                class="counter__input"
                min="0"
                value="5"
              />
              <div
                class="counter__plus"
              >
                <i
                  class="icon-plus"
                  data-testid="icon"
                />
              </div>
            </div>
            <div
              class="card__icon"
            >
              <i
                class="icon-cart"
                data-testid="icon"
              />
            </div>
          </div>
        </div>
        <div
          class="card__product--popular"
          data-testid="card-product-list"
        >
          <img
            alt="This is beef-steak images"
            class="card__image"
            src="/src/assets/images/beef-steak.png"
          />
          <div
            class="card__popular"
          >
            <a
              class="cart__link"
              href="/products/1646654668992646547"
            >
              <p
                class="text--normal"
                data-testid="text"
              >
                Beef-Steak
              </p>
            </a>
            <p
              class="price__original"
              data-testid="price"
            >
              <span
                class="price--unit"
              >
                $
              </span>
              50
            </p>
          </div>
          <div
            class="card__counter"
          >
            <div
              class="counter"
              data-testid="counter"
            >
              <i
                class="fa fa-minus"
              />
              <input
                class="counter__input"
                min="0"
                value="9"
              />
              <div
                class="counter__plus"
              >
                <i
                  class="icon-plus"
                  data-testid="icon"
                />
              </div>
            </div>
            <div
              class="card__icon"
            >
              <i
                class="icon-cart"
                data-testid="icon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
